cmd/notation/cert_gen.go

```diff
@@ -0,0 +1,139 @@
+	if len(hosts) == 0 {
+		return errors.New("missing certificate hosts")
+	}
+	name := ctx.String("name")
+	if name == "" {
+		name = hosts[0]
+	}
+
+	// generate RSA private key
+	bits := ctx.Int("bits")
+	fmt.Println("generating RSA Key with", bits, "bits")
+	key, keyBytes, err := generateTestKey(bits)
+	if err != nil {
+		return err
+	}
+
+	// generate self-signed certificate
+	cert, certBytes, err := generateTestSelfSignedCert(key, hosts, ctx.Duration("expiry"))
+	if err != nil {
+		return err
+	}
+	fmt.Println("generated certificates expiring on", cert.NotAfter.Format(time.RFC3339))
+
+	// write private key
+	keyPath := config.KeyPath(name)
+	if err := os.WriteFileWithPermission(keyPath, keyBytes, 0600, false); err != nil {
+		return fmt.Errorf("failed to write key file: %v", err)
+	}
+	fmt.Println("wrote key:", keyPath)
+
+	// write self-signed certificate
+	certPath := config.CertificatePath(name)
+	if err := os.WriteFileWithPermission(certPath, certBytes, 0644, false); err != nil {
+		return fmt.Errorf("failed to write certificate file: %v", err)
+	}
+	fmt.Println("wrote certificate:", certPath)
+
+	// update config
+	cfg, err := config.LoadOrDefault()
+	if err != nil {
+		return err
+	}
+	isDefaultKey, err := addKeyCore(cfg, name, keyPath, ctx.Bool(keyDefaultFlag.Name))
+	if err != nil {
+		return err
+	}
+	trust := ctx.Bool("trust")
+	if trust {
+		if err := addCertCore(cfg, name, certPath); err != nil {
+			return err
+		}
+	}
+	if err := cfg.Save(); err != nil {
+		return err
+	}
+
+	// write out
+	fmt.Printf("%s: added to the key list\n", name)
+	if isDefaultKey {
+		fmt.Printf("%s: marked as default\n", name)
+	}
+	if trust {
+		fmt.Printf("%s: added to the certificate list\n", name)
+	}
+	return nil
+}
+
+func generateTestKey(bits int) (crypto.Signer, []byte, error) {
+	key, err := rsa.GenerateKey(rand.Reader, bits)
+	if err != nil {
+		return nil, nil, err
+	}
+	keyBytes, err := x509.MarshalPKCS8PrivateKey(key)
+	if err != nil {
+		return nil, nil, err
+	}
+	keyPEM := pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: keyBytes})
+	return key, keyPEM, nil
+}
+
+func generateTestSelfSignedCert(key crypto.Signer, hosts []string, expiry time.Duration) (*x509.Certificate, []byte, error) {
+	now := time.Now()
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to generate serial number: %v", err)
+	}
+	template := x509.Certificate{
+		SerialNumber: serialNumber,
+		Subject: pkix.Name{
+			CommonName: hosts[0],
+		},
+		NotBefore:             now,
+		NotAfter:              now.Add(expiry),
+		KeyUsage:              x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageCodeSigning},
+		BasicConstraintsValid: true,
+	}
+	for _, host := range hosts {
+		if ip := net.ParseIP(host); ip != nil {
+			template.IPAddresses = append(template.IPAddresses, ip)
```

Comment:
curious is it required or common to add ipaddress on the self-signed certificate.

Reply:
No common but possible.